h2o-core/src/main/java/water/FailedNodeWatchdogExtension.java

```diff
@@ -0,0 +1,166 @@
+                System.arraycopy(args, i + 2, new_args, i, args.length - (i + 2));
+                return new_args;
+            }
+        }
+        return args;
+    }
+    @Override
+    public String[] parseArguments(String[] args) {
+        return parseClient(parseTimeout(args));
+    }
+
+
+    public void validateArguments() {
+        if (watchdogClientRetryTimeout < 0) {
+            H2O.parseFailed("Watchdog client retry timeout has to be positive: " + watchdogClientRetryTimeout);
+        }
+    }
+
+    @Override
+    public void onLocalNodeStarted() {
+        new FailedNodeWatchdogThread().start();
+        H2O.SELF._heartbeat._watchdog_client = watchDogClient;
+    }
+
+    /**
+     * This method checks whether the client is disconnected from this node due to some problem such as client or network
+     * is unreachable.
+     */
+    private static void handleClientDisconnect(H2ONode node, long watchdogClientRetryTimeout) {
+        if(node._heartbeat._watchdog_client){
+            Log.warn("Watchdog client " + node + " disconnected!");
+            WatchdogClientDisconnectedTask tsk = new WatchdogClientDisconnectedTask(node, watchdogClientRetryTimeout);
+            Log.warn("Asking the rest of the nodes in the cloud whether watchdog client is really gone.");
+            if((tsk.doAllNodes()).clientDisconnectedConsensus) {
+                Log.fatal("Stopping H2O cloud since the watchdog client is disconnected from all nodes in the cluster!");
+                H2O.shutdown(0);
+            }
+        }else if(node._heartbeat._client) {
+            Log.warn("Client "+ node +" disconnected!");
+        }
+
+        // in both cases remove the client since the timeout is out
+        if(node._heartbeat._client){
+            H2O.removeClient(node);
+            if(H2O.isFlatfileEnabled()){
+                H2O.removeNodeFromFlatfile(node);
+            }
+        }
+    }
+
+    /**
+     * Helper MR task used to detect clientDisconnectedConsensus on the timeout we last heard from the watchdog client
+     */
+    private static class WatchdogClientDisconnectedTask extends MRTask<WatchdogClientDisconnectedTask> {
+        private boolean clientDisconnectedConsensus = false;
+        private H2ONode clientNode;
+        private long watchdogClientRetryTimeout;
+        WatchdogClientDisconnectedTask(H2ONode clientNode, long  watchdogClientRetryTimeout) {
+            this.clientNode = clientNode;
+            this.watchdogClientRetryTimeout = watchdogClientRetryTimeout;
+        }
+
+        @Override
+        public void reduce(WatchdogClientDisconnectedTask mrt) {
+            this.clientDisconnectedConsensus = this.clientDisconnectedConsensus && mrt.clientDisconnectedConsensus;
+        }
+
+        @Override
+        protected void setupLocal() {
+            H2ONode desiredClient = null;
+            for (H2ONode client : H2O.getClients()) {
+                if (client.equals(clientNode)) {
```

Comment:
I can check if the client is in the list to prevent from iterating over the list when the client is not there, however I can't get it from the hashset right away. I could change the return type to something else, but wanted to be consistent with the rest of the methods